### PR TITLE
feat: add property prediction module with pretrained ROOST band gap model

### LINKED ISSUE #448: WIP feat: SMACT Property Prediction Module
This module integrates property prediction capabilities directly into the SMACT codebase, utilising trained ROOST model checkpoints. It enables users to predict material properties directly from chemical compositions, without requiring crystal structure information.

This integration is part of ongoing development initiated during the 2025 SMACTDOWN session. Once fully integrated, users will be able to compute target material properties such as band gap or bulk modulus through the SMACT interface.

## Module Structure

```text
smact/property_prediction/
├── __init__.py              # Main module interface
├── base_predictor.py        # Abstract base class
├── wrapper.py           # Convenience functions (i.e. predict_bandgap(model, composition)) default Roost
├── roost/
│   ├── __init__.py         # ROOST module
│   └── predictor.py        # ROOST implementation
└── example_usage.py        # Example usage demonstration
```

## Proposed Implementation

The proposed usage pattern would be:

```python
from smact.property_prediction import RoostPropertyPredictor

# Check available properties
print(RoostPropertyPredictor.available_properties)  # ['band_gap', 'bulk_modulus']

# Create model with just property name and device
model = RoostPropertyPredictor(property_name="band_gap", device="cpu")

# Predict single or multiple compositions
model.predict("NaCl")                    # Single composition
model.predict(["NaCl", "TiO2", "GaN"])   # Multiple compositions
```

## Key Features

1. **Class-level properties**: `RoostPropertyPredictor.available_properties`
2. **SMACT validation**: All compositions validated using `smact_validity()`
3. **Extensible design**: Easy to add new property prediction model types 
6. **Error handling**: Clear error messages for invalid compositions/properties

## Extensibility for Multiple Models

The architecture is designed for multiple models:

```python
# Future implementations will work the same way:
from smact.property_prediction import OtherPropertyPredictor

other_model = OtherPropertyPredictor(property_name="bulk_modulus")
```

## Abstract Base Class Architecture

In `/smact/property_prediction/base_predictor.py`:

```python
from abc import ABC, abstractmethod

class BasePropertyPredictor(ABC):  # Abstract base class
    """Abstract base class for property predictors."""
    
    @property
    @abstractmethod
    def supported_properties(self) -> list[str]:  # Must be implemented
        """List of properties supported by this predictor."""
        pass
    
    @abstractmethod
    def predict(self, compositions: str | list[str]) -> np.ndarray:  # Must be implemented
        """Predict property values for given compositions."""
        pass
```

## Concrete Implementation

The `RoostPropertyPredictor` **inherits from and implements** the abstract base class:

```python
class RoostPropertyPredictor(BasePropertyPredictor):  # Concrete implementation
    
    @property
    def supported_properties(self) -> list[str]:  # Implements abstract method
        return self.available_properties
    
    def predict(self, compositions: str | list[str]) -> np.ndarray:  # Implements abstract method
        compositions = self._validate_compositions(compositions)
        # Implementation here...
        return np.zeros(len(compositions))  # Placeholder for now
```

## Extensible Pattern

This allows for multiple concrete implementations:

```python
# Future models will also inherit from BasePropertyPredictor
class OtherPropertyPredictor(BasePropertyPredictor):
    def supported_properties(self) -> list[str]:
        return ["band_gap", "formation_energy"]
    
    def predict(self, compositions: str | list[str]) -> np.ndarray:
        # Other-specific implementation
        pass
```

## Current Status
The module is ready for integration with actual ROOST model checkpoints once training is complete.

### Contributors
Hyunsoo Park
Jamie Swaine
Matthew Walker
Yuan Li
Zibo Zhou
Ry Nduma

### Acknowledgments
With thanks to the broader SMACT community, and especially to participants of SMACTDOWN 2025 for their feedback and contributions.

### Citations:

`roost` - Predicting materials properties without crystal structure: Deep representation learning from stoichiometry. [[Paper]](https://doi.org/10.1038/s41467-020-19964-7) [[arXiv]](https://arxiv.org/abs/1910.00617)

```bibtex
@article{goodall_2020_predicting,
  title={Predicting materials properties without crystal structure: Deep representation learning from stoichiometry},
  author={Goodall, Rhys EA and Lee, Alpha A},
  journal={Nature Communications},
  volume={11},
  number={1},
  pages={1--9},
  year={2020},
  publisher={Nature Publishing Group}
}
```

### PR DESCRIPTION
**Pull request by ryannduma**
## Summary

Adds the `smact.property_prediction` module — a composition-only property prediction framework using pretrained ROOST (Representation Learning from Stoichiometry) models. This enables users to predict material properties like band gap from a chemical formula string alone, with no crystal structure required.

Closes #448

## What's new

- **`smact.property_prediction` subpackage** — full module with ABC-based predictor interface, model I/O, registry, and config
- **`predict_band_gap()` convenience function** — one-liner prediction from composition strings
- **`RoostPropertyPredictor` class** — flexible predictor with uncertainty quantification via heteroscedastic (robust) loss
- **`pretrained_models/` directory** — MatGL-style model versioning with bundled ROOST checkpoint (band gap, 0.28 eV MAE on 103K MP DFT samples)
- **Sphinx documentation** — 7 new `.rst` pages covering architecture, API reference, model performance, training guide, and known limitations
- **31 unit tests** in `smact/property_prediction/tests/` with `pytest.importorskip` guards for optional dependencies

### Architecture

The module follows an **Abstract Base Class pattern** (`BasePropertyPredictor` → `RoostPropertyPredictor`) to support future model backends (e.g., CGCNN, MEGNet) without API changes. All predictors share:
- `predict(compositions, return_uncertainty=False)` → `np.ndarray | PredictionResult`
- `PredictionResult` dataclass with `.predictions`, `.uncertainties`, `.epistemic_std`, `.to_dict()`

### Model resolution chain (inspired by MatGL)

When a model is requested, the loader checks in order:
1. **Local directory path** — user-supplied explicit path
2. **`pretrained_models/`** — bundled in the repo (for source installs / development)
3. **`~/.cache/smact/models/`** — downloaded models cached locally
4. **Remote download** — fetches `.tar.gz` from GitHub Releases

This means `predict_band_gap("NaCl")` works out-of-the-box when running from the repo, with no manual path configuration.

### Pretrained models directory (MatGL pattern)

Following [AntObi's recommendation](https://github.com/WMD-group/SMACT/issues/448) and [MatGL's approach](https://github.com/materialsvirtuallab/matgl/tree/main/pretrained_models), each model lives in its own versioned directory with:
- `model.json` — metadata and hyperparameters
- `model.pt` — model architecture parameters + normaliser dict
- `state.pt` — trained weights
- `README.md` — per-model card (aim, training data, performance, references)

Models are **not distributed via PyPI** (too large). They're available when cloning the repo or via the download/cache mechanism.

### Model naming convention

```
Roost-<dataset>-<version>-<property>[-<fidelity>]
```
Example: `Roost-MP-2024.12.0-band_gap` (ROOST trained on Materials Project, version 2024.12.0, predicting band gap)

### Optional dependency handling and lazy imports

`torch` and `aviary-models` are behind the `property_prediction` extra in `pyproject.toml`:

```toml
property_prediction = ["aviary-models>=1.2.0", "torch>=2.0.0"]
```

The rest of SMACT works without these heavy dependencies. To ensure the package is importable *without* torch installed, all torch/requests/aviary imports use **lazy loading**:

- **`__init__.py` and `roost/__init__.py`** use `__getattr__()` + `TYPE_CHECKING` imports so that `RoostPropertyPredictor` and `predict_band_gap` are only resolved when actually accessed, not when the package is first imported.
- **`io.py` and `registry.py`** defer `import torch` and `import requests` to function bodies (`load_checkpoint()`, `save_checkpoint()`, `get_available_models()`, etc.) rather than module level.
- **`convenience.py`** defers `from smact.property_prediction.roost import RoostPropertyPredictor` into the function body.

This matters because without it, the import chain `smact.property_prediction` → `roost/__init__.py` → `predictor.py` → `import torch` would crash during pytest collection before `pytest.importorskip("torch")` ever ran. The lazy pattern means:

- `from smact.property_prediction import get_supported_properties` works without torch
- `from smact.property_prediction import BasePropertyPredictor` works without torch
- `pytest.importorskip("torch")` can run and gracefully skip torch-dependent tests
- `from smact.property_prediction import predict_band_gap` works (creates the function object), but *calling* it triggers the torch import

Tests use `pytest.importorskip()` to skip gracefully in environments without torch.

## Testing

- **31 tests** across `test_base_predictor.py`, `test_io.py`, and `test_registry.py`
- Tests added to `testpaths` in `pyproject.toml` so CI discovers them
- `pytest.importorskip("torch")` and `pytest.importorskip("requests")` guard tests requiring optional deps
- CI install updated: `uv sync --extra optional --extra property_prediction --dev`
- All tests pass locally; `ruff check`, `pyright`, `codespell`, and `pre-commit` all clean

### Coverage

CI installs `--extra property_prediction` so torch/aviary are available and IO tests run. Unit tests cover the config, registry, base predictor, and IO layers. The ROOST predictor, training pipeline, convenience wrapper, and checkpoint converter currently have 0% coverage — exercising them requires loading the 9MB model and running inference, which is integration-test territory. Codecov is configured with `fail_ci_if_error: False` so it will report but not block.

### Pyright: remaining `reportMissingImports`

After this PR, `uv run pyright smact/` reports **18 errors, 0 warnings** — down from 20 errors before. All 18 are `reportMissingImports` for optional dependencies not installed in the dev/lint environment:

| Package                  | Files                                                        | Count | Status       |
|--------------------------|--------------------------------------------------------------|-------|--------------|
| `mp_api.client`          | database.py, structure.py, download_compounds_with_mp_api.py | 3     | Pre-existing |
| `torch` / `torch.utils.data` | predictor.py, train.py, io.py, convert_checkpoint.py    | 7     | Optional dep |
| `aviary.*`               | predictor.py, train.py                                       | 7     | Optional dep |
| `sklearn.model_selection` | train.py                                                    | 1     | Optional dep |

These are **not bugs** — they resolve correctly when `pip install smact[property_prediction]` is used. CI installs `--extra property_prediction` so they resolve there. We chose not to suppress them because:

- Setting `reportMissingImports = false` in pyproject.toml would mask real missing-import bugs across the entire codebase
- Per-line `# pyright: ignore` comments add noise to every import
- Moving them into `TYPE_CHECKING` blocks would break runtime code — `predictor.py` and `train.py` actually use torch/aviary at runtime
- The 3 pre-existing `mp_api` errors prove the SMACT team already accepts this pattern for optional deps
- The only clean fix is installing all optional deps in the pyright pre-commit environment, which is a broader CI/config decision

## Test plan

- [x] `uv run pytest smact/property_prediction/tests/ -v` — 30 passed, 1 skipped without torch
- [x] `uv run pre-commit run --all-files` — all hooks pass on our code
- [x] `uv run sphinx-build -b html docs docs/_build/html` — docs build successfully
- [x] Verify `from smact.property_prediction import get_supported_properties` works without torch installed

## Documentation

- `docs/smact.property_prediction.rst` — main docs page with architecture overview, ROOST data flow, model performance table, baseline comparisons, known limitations, model versioning guide, training guide, and references
- 6 submodule `.rst` files for autodoc (`base_predictor`, `config`, `convenience`, `io`, `registry`, `roost`)
- `docs/conf.py` updated with `autodoc_mock_imports` for `torch`, `aviary`, `requests`
- `docs/index.rst` updated with property prediction in features list
- `docs/smact.rst` updated with `smact.property_prediction` in toctree
- `pretrained_models/README.md` — top-level model index
- `pretrained_models/Roost-MP-2024.12.0-band_gap/README.md` — per-model card

### Quick start

```python
from smact.property_prediction import predict_band_gap

# Single composition
result = predict_band_gap("GaN")
# array([1.59])

# Multiple compositions
results = predict_band_gap(["NaCl", "TiO2", "GaN", "Si"])
# array([5.67, 2.13, 1.59, 0.72])
```

### With uncertainty

```python
from smact.property_prediction import RoostPropertyPredictor

predictor = RoostPropertyPredictor(property_name="band_gap")
result = predictor.predict(["NaCl", "TiO2"], return_uncertainty=True)

print(result.predictions)     # [5.67, 2.13]
print(result.uncertainties)   # [0.42, 0.31]  (aleatoric)
print(result.unit)            # "eV"
```

### Registry queries

```python
from smact.property_prediction import get_supported_properties, get_available_models

get_supported_properties()  # ["band_gap"]
get_available_models()      # ["Roost-MP-2024.12.0-band_gap"]
```

### Tutorial notebook

A new Jupyter notebook at `docs/examples/property_prediction.ipynb` provides a hands-on walkthrough:

1. **Quick start** — `predict_band_gap("GaN")` one-liner
2. **RoostPropertyPredictor class** — loading a model once and reusing across calls
3. **Uncertainty quantification** — `return_uncertainty=True` for aleatoric uncertainties from robust loss
4. **Model registry** — querying supported properties, available models, and units
5. **SMACT screening + prediction** — using `smact_filter` to generate candidate compositions, then predicting band gaps for all candidates in one batch

The notebook is included in the Sphinx docs via `myst-nb` and linked from the Examples toctree.

## Model performance

| Property | Model                       | Dataset           | MAE     | R²   | N samples |
|----------|-----------------------------|-------------------|---------|------|-----------|
| Band gap | Roost-MP-2024.12.0-band_gap | Materials Project | 0.28 eV | 0.93 | 103,038   |

### New files

- `smact/property_prediction/` — full subpackage (7 modules + `__init__.py`)
- `smact/property_prediction/tests/` — 31 unit tests
- `pretrained_models/` — bundled model checkpoint + READMEs
- `docs/smact.property_prediction*.rst` — 7 documentation pages
- `docs/examples/property_prediction.ipynb` — tutorial notebook

### Modified files

- `pyproject.toml` — `property_prediction` extra, testpaths, ruff per-file-ignores
- `.github/workflows/ci.yml` — `--extra property_prediction` for CI coverage
- `docs/conf.py` — `autodoc_mock_imports`
- `docs/index.rst` — features list
- `docs/smact.rst` — toctree
- `docs/examples.rst` — toctree

## Summary by CodeRabbit

* **New Features**
  * Property prediction from composition using pre-trained ROOST models (band gap) with optional uncertainty, registry, local cache/remote retrieval, convenience API, training/conversion tooling, and an included pretrained model.

* **Documentation**
  * Extensive new docs, examples and notebook demonstrating property prediction and model catalogue; docs year updated to 2026; pretrained models README added.

* **Tests**
  * New unit and integration tests covering predictors, I/O and registry.

* **Chores**
  * CI lint/test invocations updated to include property_prediction flag; .gitignore extended.

### CODE REVIEW

File: smact/property_prediction/io.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+                f.write(chunk)
+
+        logger.info(f"Downloaded to {tar_path}, extracting...")
+
+        # Extract the archive safely using data filter for security
+        with tarfile.open(tar_path, "r:gz") as tar:
+            if sys.version_info >= (3, 11, 4):
+                tar.extractall(self.cache_location, filter="data")
+            else:
+                tar.extractall(self.cache_location)  # noqa: S202
```
[the remark this answers is not in the document]
Comment (ryannduma) [reply smact/property_prediction/io.py]: io.py — tar filter CVE version bump to 3.11.13 — The bot's web search is wrong. filter="data" was added in 3.11.4 and works correctly.
  CVE-2025-4517 is about a filter bypass in contrived scenarios — we're extracting from our own known-good model archives, not untrusted input.
  Not applicable.

**Review by ryannduma — COMMENTED** [review]
Ive added 3 commits addressing Copilot and CodeRabbit review comments across the property prediction module.
                                         
  Commit 1: 5be15e2 — Address remaining bot review comments

  - Notebook (property_prediction.ipynb): Added strict=True to 3 zip() calls, renamed unused charges → _charges
  - predictor.py: Added warning for checkpoint target mismatch; added warning when uncertainty requested on non-robust model
  - train.py: Added validation for test_size/val_size split fractions
  - test_base_predictor.py: Narrowed except Exception → except (ImportError, ModuleNotFoundError)
  - test_io.py: Mocked RemoteFile to prevent network access in tests
  - io.py: Added Python 3.11.4 version guard for tar.extractall(filter="data")
  - README.md: Replaced hardcoded /home/ryan/... path with relative path
  - registry.py: Caught json.JSONDecodeError in manifest fetch; added _is_valid_model_dir helper; added allow_redirects=True to requests.head()
  - pyproject.toml: Bumped aviary-models>=1.2.1

  Commit 2: 5684831 — Fix bot review follow-ups

  - predictor.py: Upgraded target mismatch from warning → ValueError (prevents unit mislabeling)
  - registry.py: _is_valid_model_dir now requires both model.pt and state.pt (matches load_model_files contract)
  - io.py: RemoteFile validates cached model completeness before skipping download; load_model_files validates files exist after download
  - pyproject.toml: Added requests to property_prediction optional dependencies

  Commit 3: a55286b — Fix format string bug

  - train.py: Fixed TypeError when best_val_loss missing from metadata (.4f format on "N/A" string)

  Intentionally not addressed

  - tar CVE version bump (io.py:85): filter="data" works correctly from 3.11.4. CVE-2025-4517 is a contrived filter bypass irrelevant to
  extracting our own model archives.
  - parse_model_name refactor (registry.py:245): Over-engineering for a theoretical edge case where a fidelity name matches a property word
  fragment.
  - Fail-fast on uncertainty for non-robust (predictor.py:223): Warning is intentional — returning PredictionResult(uncertainties=None) is useful
  and doesn't break downstream code. A ValueError would be a breaking API change.
  - Dataset-size feasibility check (train.py:142): train_test_split already raises a clear error for empty splits.

  All CI checks pass (lint, type check, tests on Python 3.11–3.13 across Ubuntu/macOS/Windows). 
  
  Happy with it's state ahead of merge :))

**Review by KingaMas — APPROVED** [review]
Reviewed the property prediction module. The code works as expected and all
  tests pass (309 passed, 2 skipped).                                             
   
  Coverage is at 89% overall. The remaining uncovered lines are in:               
  - Plotting/CLI code (plot_embedding.py, band_gap_simple.py) — not unit-testable
  - Benchmarking scripts — profiling utilities, not library code
  - Exception branches inside tests that only trigger on test failure
  - MP API tests gated behind API key availability

  These are excluded from coverage measurement for good reason.

  Merging.
